Android backend (input and output)

ligfx opened the pull request:
A.K.A. make libsoundio work on all major OSes (does the CoreAudio backend work on iOS?)
A.K.A. better than SDL
A.K.A. 500 lines and some hard work gets you surprisingly far
A.K.A. The Cubs are still going to win the World Series, I just know it

Couple points:
- Android has a special audio state it calls the "fast track," which skips software mixing and ties to reduce latency as much as possible. The keys to this are matching the device's natural sample rate and buffer size.

  Unfortunately, there's no officially supported way to get this information outside of a JVM running inside an Android Activity. There are varying degrees of hacks around this, but they could be disappeared anytime or, worse, crash the entire process (ex. calling `JNI_GetCreatedJavaVMs` on Android without a pre-existing JVM raises an assertion error, so you can't easily opportunistically attach to a JVM. Even if you did get around it, there's still no officially supported way to get an Activity handle from the global context.)

  In an ideal world, the backend would default to the fastest sample rate and a proper multiple of buffer size. Failing that, sample rate can be set manually and buffer size can be requested proportional to latency and sample rate.

- I only implemented support for output devices. If that's a deal-breaker, I can go read-up on how input devices work on Android. They have some quirks that make them different from output devices and might possibly need JVM access (with the problems inherent in that, as described above).

- This backend does not support calling `soundio_outstream_begin_write/soundio_outstream_end_write` multiple times in a single write callback. This part of the libsoundio API has always seemed weird to me, and it doesn't map well onto OpenSL ES. If you do call the functions more than once, you'll get an error because the backend buffers are full.

- No underflow callbacks. I couldn't figure this one out. There's an interface called `SLPrefetchStatusItf` that looks like it could be handy here, but in local testing it only raised events when a write callback was not in the process of being called, which made it pretty much useless (since by the time the callback ends, the underflow state has been "fixed"). It also seems to randomly cause assertion errors when used with Android's buffer queue implementation, and nobody on the Internet knows why.

  I thought about implementing an additional thread to intercept write callbacks from OpenSL ES, but then you'd lose out on the special thread priority that audio callbacks get and it complicates the code.

- Requires API level 21 / Android 5.0 Lollipop or greater. The main feature this brings is floating-point support. The rest of the codebase doesn't work with API levels lower than that anyways.

- Android doesn't have a `/tmp` folder, but it does follow the POSIX convention of pointing the TMPDIR environment variable to an appropriate location. I just added this location to `os.c` for the sake of simplicity, though ideally it would read the right environment variables.

- Testing this out is a little hairy depending on your system, but something like this plopped in at the top of `CMakeLists.txt` (before `project(...)`) did the trick for me:

  ```cmake
set(BUILD_DYNAMIC_LIBS OFF CACHE BOOL "" FORCE)
set(CMAKE_SYSROOT /usr/local/Cellar/android-ndk/r13/platforms/android-24/arch-arm)
set(CMAKE_C_COMPILER /usr/local/Cellar/android-ndk/r13/toolchains/arm-linux-androideabi-4.9/prebuilt/darwin-x86_64/bin/arm-linux-androideabi-gcc)
set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fPIC")
set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -pie")
```

   Follow that up with a

   ```bash
adb push sio_sine /data/local/tmp && adb shell /data/local/tmp/sio_sine
```

   and you'll hear the sweet dulcet tones of a sine wave coming from your device's speakers.

- Overall, the libsoundio API was a pleasure to work with! It's verbose in many places, but I never ran into any problems mapping OpenSL ES onto the abstraction model or figuring out where things belonged.

- I took a look at the v2 branch, I think the only change that affect this is the `int` to `enum SoundIoError` change in function prototypes.

EDIT: Added support for input devices.

- Supposedly, not all devices support all sample rates. However, on my LG G4 all sample rates seem to be supported. I'm not sure what will happen if you try to open a device with an unsupported sample rate. Anecdotal evidence on the internet says sometimes it immediately fails and sometimes it just returns corrupted audio.

- Is it useful to know if a device is currently unavailable due to permissions problems? I'm leaning towards yes, though it's not straightforward to figure out from the backend. On my device, the error code for lacking permissions was SL_RESULT_CONTENT_UNSUPPORTED, but according to the spec it should probably be SL_RESULT_PERMISSION_DENIED.

- Android lets you make requests for different recording profiles: GENERIC, CAMCORDER, VOICE_COMMUNICATION, and VOICE_RECOGNITION. The camcorder request could be exposed as a separate device (and probably should, though it's not guaranteed to actually be different), but what about the VOICE_* options? They might be too high-level for this project, and more appropriate for an effects layer.

- Recording floating-point data is supported starting in API level 23 / Android 6.0 Marshmallow. I put a check into the backend init function to try opening a device in that format, which seems to work (tested on an emulator running API level 21).

- Recording is much more difficult to test, because command-line processes don't have permissions to access input devices. To try this out I ended up creating a [NativeActivity](https://developer.android.com/reference/android/app/NativeActivity.html) with copy-pasted code from `sio_record`. This worked on my real device, but not on the emulator because I couldn't figure out where I had permission to write files. Should there be a small sample app for testing (probably similar to `sio_microphone`), or is that too complicated?

fennecdjay commented:
Looks great!
but I must admit I'd love to see input handling implemented

andrewrk commented on src/android.c:
```diff
@@ -0,0 +1,402 @@
+
+    // TODO: An audio stream's latency and jitter can be reduced by using the
+    // native buffer size. Unfortunately, there is no officially supported way
+    // to get the native buffer size without access to a JVM.
```
Hmm, maybe we can expose a way to override the buffer size, and then in the user's code if they somehow have access to the JVM, then they can pass in this buffer size.

ligfx replied:
I've been getting around this by requesting a latency of `FACTOR OF BUFFER SIZE / SAMPLING RATE`, which seems to work.

It would've been nice to set the default latency to something that results in the native buffer size, though.

andrewrk commented on src/android.c:
```diff
@@ -0,0 +1,402 @@
+    osa->bytes_per_buffer = outstream->bytes_per_frame * outstream->sample_rate
+        * outstream->software_latency / 2;
+    size_t samples_per_buffer = osa->bytes_per_buffer / outstream->bytes_per_sample;
+    osa->buffers[0] = calloc(samples_per_buffer, outstream->bytes_per_sample);
```
Don't forget to check for memory allocation failures!

andrewrk commented on src/android.c:
```diff
@@ -0,0 +1,402 @@
+    if (SL_RESULT_SUCCESS != (*osa->playerPlay)->SetPlayState(osa->playerPlay,
+        pause ? SL_PLAYSTATE_PAUSED : SL_PLAYSTATE_PLAYING))
+    {
+        return SoundIoErrorInvalid;
```
`SoundIoErrorInvalid` I believe is meant to mean "you called this function with invalid state" such as passing NULL for `si`. If SetPlayState returns an error, I think we want one of these:

```
    /// An open stream had an error that can only be recovered from by
    /// destroying the stream and creating it again.
    SoundIoErrorStreaming,

    /// Attempted to use a device with parameters it cannot support.
    SoundIoErrorIncompatibleDevice,

    /// Attempted to use parameters that the backend cannot support.
    SoundIoErrorIncompatibleBackend,
```

ligfx replied:
The other return values of `SetPlayState` are:
- `SL_RESULT_CONTENT_CORRUPTED`: definitely `SoundIoErrorStreaming`
- `SL_RESULT_PARAMETER_INVALID`: this should never happen, but `SoundIoErrorIncompatibleDevice` I guess?
- `SL_RESULT_PERMISSION_DENIED` and `SL_RESULT_CONTENT_UNSUPPORTED`: the first doesn't seem to be used on Android, but the second is used for the same idea: the process doesn't have authorization to record audio. Maybe this is a new error type?

andrewrk commented:
One thing that I'm not clear on - I thought that to write an Android application, you had to write Java? But libsoundio is C. How does that work?


> This backend does not support calling soundio_outstream_begin_write/soundio_outstream_end_write multiple times in a single write callback. This part of the libsoundio API has always seemed weird to me, and it doesn't map well onto OpenSL ES. If you do call the functions more than once, you'll get an error because the backend buffers are full.

You don't have to make that possible. Most backends only need one write operation per callback. However the ALSA API does some kind of memory mapping strategy which can possibly require multiple write operations per callback which is why we have that API.

> Requires API level 21 / Android 5.0 Lollipop or greater. The main feature this brings is floating-point support. The rest of the codebase doesn't work with API levels lower than that anyways.

Somewhat related, I wonder if in a major version update we could eliminate floating point entirely from libsoundio. Maybe we don't need it.



> Android doesn't have a /tmp folder, but it does follow the POSIX convention of pointing the TMPDIR environment variable to an appropriate location. I just added this location to os.c for the sake of simplicity, though ideally it would read the right environment variables.

Good point. I created #128

> Overall, the libsoundio API was a pleasure to work with! It's verbose in many places, but I never ran into any problems mapping OpenSL ES onto the abstraction model or figuring out where things belonged.

Glad to hear that.

> I took a look at the v2 branch, I think the only change that affect this is the int to enum SoundIoError change in function prototypes.

I think the only thing that breaks compatibility with this PR is adding the SoundIoBackend to the 2nd to last enum position instead of the last position. I think I should move the position of SoundIoBackendDummy to not be the last in v2 so that we can avoid this problem in the future.

Regardless, I think that we will release this in v2 instead of a minor version, because more releases is more work, and I'd rather push forward into the future than enable people to be stuck on old versions of stuff.

Yepoleb commented:
Good support for input devices on all platforms is the only reason I went with libsoundio instead of using platform specific backends. I'd hate to see this major advantage go. Output only is fine for a preliminary implementation, but the release version should follow the tradition of supporting all APIs equally.

andrewrk commented:
> Output only is fine for a preliminary implementation, but the release version should follow the tradition of supporting all APIs equally

I think I agree with this.

iskunk commented:
> Somewhat related, I wonder if in a major version update we could eliminate floating point entirely from libsoundio. Maybe we don't need it.

I thought floating-point was favored in some high-end sound applications, as a way of avoiding clipping (you have all-but-infinite headroom). Scientific applications, too. What benefit would there be to getting rid of it?

ligfx commented:
> One thing that I'm not clear on - I thought that to write an Android application, you had to write Java? But libsoundio is C. How does that work?

You package a shared library inside your app, which gets loaded into the JVM and fulfills specially-marked JNI functions, similar to writing a C extension for any other language. Android provides the NDK specifically for this. For example, if you disassemble a Netflix app, a number of DRM libraries are included in binary form.

OpenSL ES is a good use case for dropping down to C, because the JVM would add latency to every callback.

> You don't have to make that possible. Most backends only need one write operation per callback. However the ALSA API does some kind of memory mapping strategy which can possibly require multiple write operations per callback which is why we have that API.

Cool.

> > Somewhat related, I wonder if in a major version update we could eliminate floating point entirely from libsoundio. Maybe we don't need it.

> I thought floating-point was favored in some high-end sound applications, as a way of avoiding clipping (you have all-but-infinite headroom). Scientific applications, too. What benefit would there be to getting rid of it?

The [Android documentation](https://developer.android.com/ndk/guides/audio/sampling-audio.html#floating-point) has this to say:

Using floating-point numbers to represent audio data can significantly enhance audio quality in high-performance audio applications. Floating point offers the following advantages:

- Wider dynamic range.
- Consistent accuracy across the dynamic range.
- More headroom to avoid clipping during intermediate calculations and transients.

> > Output only is fine for a preliminary implementation, but the release version should follow the tradition of supporting all APIs equally

> I think I agree with this.

I'll take a look, this will take me longer to figure out how to get proper permissions.

ligfx commented:
> Regardless, I think that we will release this in v2 instead of a minor version, because more releases is more work, and I'd rather push forward into the future than enable people to be stuck on old versions of stuff.

Cool, I can rebase.

> I think the only thing that breaks compatibility with this PR is adding the SoundIoBackend to the 2nd to last enum position instead of the last position. I think I should move the position of SoundIoBackendDummy to not be the last in v2 so that we can avoid this problem in the future.

I don't understand. What do you mean?

ligfx commented:
Okay, I added support for input devices. I added some notes to the PR summary, which I've copied here:

- Same issue with native sample rates and buffer sizes.

- Supposedly, not all devices support all sample rates. However, on my LG G4 all sample rates seem to be supported. I'm not sure what will happen if you try to open a device with an unsupported sample rate. Anecdotal evidence on the internet says sometimes it immediately fails and sometimes it just returns corrupted audio.

- Is it useful to know if a device is currently unavailable due to permissions problems? I'm leaning towards yes, though it's not straightforward to figure out from the backend. On my device, the error code for lacking permissions was SL_RESULT_CONTENT_UNSUPPORTED, but according to the spec it should probably be SL_RESULT_PERMISSION_DENIED.

- Android lets you make requests for different recording profiles: GENERIC, CAMCORDER, VOICE_COMMUNICATION, and VOICE_RECOGNITION. The camcorder request could be exposed as a separate device (and probably should), but what about the VOICE_* options? They might be too high-level for this project, and more appropriate for an effects layer.

- Recording floating-point data is supported starting in API level 23 / Android 6.0 Marshmallow. I put a check into the backend init function to try opening an object in that format, which seems to work (tested on an emulator running API level 21).

- Recording is much more difficult to test, because command-line processes don't have permissions to access input devices. To try this out I ended up creating a [NativeActivity](https://developer.android.com/reference/android/app/NativeActivity.html) with copy-pasted code from `sio_record`. This worked on my real device, but not on the emulator because I couldn't figure out where I had permission to write files. Should there be a small sample app for testing (probably similar to `sio_microphone`), or is that too complicated?

andrewrk commented:
> Is it useful to know if a device is currently unavailable due to permissions problems? I'm leaning towards yes, though it's not straightforward to figure out from the backend. On my device, the error code for lacking permissions was SL_RESULT_CONTENT_UNSUPPORTED, but according to the spec it should probably be SL_RESULT_PERMISSION_DENIED.

Sometimes we learn that devices are unavailable during device scanning. That's useful information but not always available. The real time that you learn whether a device could be opened is when you try to open it for input or output. So I think the way you have it is fine.

Seems like for Android it might be a little more complicated, because doesn't it potentially prompt the user for permission when an application attempts to play sound or use the microphone for the first time? I'm not sure how that would work.

> Android lets you make requests for different recording profiles: GENERIC, CAMCORDER, VOICE_COMMUNICATION, and VOICE_RECOGNITION. The camcorder request could be exposed as a separate device (and probably should, though it's not guaranteed to actually be different), but what about the VOICE_* options? They might be too high-level for this project, and more appropriate for an effects layer.

Maybe we can introduce some API for this. Something like `soundio_device_get_purpose()` and it might return `SoundIoErrorUnsupportedBackend` or it might return a `SoundIoDevicePurposeFoo` enum value which would correspond to GENERIC, CAMCORDER, etc. In general, the philosophy is to give the API all the power of all the backends, abstract over the functionality that makes sense to do that for, document it well, and the libsoundio API consumer can choose to ignore the backend-specific functionality or choose to take advantage of it.

> Using floating-point numbers to represent audio data can significantly enhance audio quality in high-performance audio applications. Floating point offers the following advantages:

> I thought floating-point was favored in some high-end sound applications, as a way of avoiding clipping (you have all-but-infinite headroom). Scientific applications, too. What benefit would there be to getting rid of it?

I was mistaken to propose eliminating floating point. Silly, let's pretend I didn't suggest that.



> > I think the only thing that breaks compatibility with this PR is adding the SoundIoBackend to the 2nd to last enum position instead of the last position. I think I should move the position of SoundIoBackendDummy to not be the last in v2 so that we can avoid this problem in the future.
> I don't understand this. What do you mean?

What I mean is that when you add an item to the middle of an enum in an .h file, the dynamic library you produce is no longer ABI compatible with software compiled against the older .h file. So when I break major compatibility, I will put the dummy item toward the beginning, so that in the future if any new backends are added they can go in the last position of the enum, preserving ABI compatibility.

> Recording is much more difficult to test, because command-line processes don't have permissions to access input devices. To try this out I ended up creating a NativeActivity with copy-pasted code from sio_record. This worked on my real device, but not on the emulator because I couldn't figure out where I had permission to write files. Should there be a small sample app for testing (probably similar to sio_microphone), or is that too complicated?

Difficulty of testing is a big problem. It's the main reason that it's taking so long to get a release out and the main reason that it's taking so long to integrate this pull request. In general, the easier we make it to test libsoundio, the higher quality it will be, and the easier it will be to contribute to.

ligfx commented:
> Difficulty of testing is a big problem. It's the main reason that it's taking so long to get a release out and the main reason that it's taking so long to integrate this pull request. In general, the easier we make it to test libsoundio, the higher quality it will be, and the easier it will be to contribute to.

Small update about testability: I was struggling with different ways people have used to integrate CMake and the Android NDK, when it was brought to my attention that the latest NDK actually includes its own CMake toolchain file. I don't see any problem with requiring the latest version of the NDK to build.

I've updated the readme accordingly, so now testing outstreams should be as easy as 1) installing the NDK and 2) copy-pasting a line of shell code. This has been tested on macOS, can anybody confirm that it works on other OSes? 

Making testing instreams easy is the next step.

> I was mistaken to propose eliminating floating point. Silly, let's pretend I didn't suggest that.

Forgotten!

> What I mean is that when you add an item to the middle of an enum in an .h file, the dynamic library you produce is no longer ABI compatible with software compiled against the older .h file. So when I break major compatibility, I will put the dummy item toward the beginning, so that in the future if any new backends are added they can go in the last position of the enum, preserving ABI compatibility.

Ah doi, I knew that. I know this isn't the place to discuss this really, but what's your reasoning for changing ABI compatibility on the enum? I don't have issues with the way it is currently (other than giving preoccupied contributors a place to screw up).

> Seems like for Android it might be a little more complicated, because doesn't it potentially prompt the user for permission when an application attempts to play sound or use the microphone for the first time? I'm not sure how that would work.

Generally, code that requires permissions will just fail (silently or not) if the app doesn't already have that permission. The suggested way to handle this is to check permissions before using associated features, which does sometimes prompt the user (starting in Marshmallow, I think?).

> Maybe we can introduce some API for this.

That makes sense to me. Down to have that be a separate issue/PR, for the sake of simplicity?

andrewrk commented:
> what's your reasoning for changing ABI compatibility on the enum? I don't have issues with the way it is currently (other than giving preoccupied contributors a place to screw up).

That's the only reason - to make it easier for future contributors. There's a cost in going from 0 ABI breakage to 1 ABI breakage, but there's no cost going from 1 ABI breakage to N. So we might as well do it all at once.


> > Maybe we can introduce some API for this.
> That makes sense to me. Down to have that be a separate issue/PR, for the sake of simplicity?

Hmm. I think it's reasonable to merge this and then solve that problem later.

andrewrk commented:
Awesome work on this. I'm merging into v2 branch right now.

andrewrk commented:
Alright, let's get some testing going and then release time :tada:
